crates/core/component/dex/src/component/position_manager.rs

```diff
@@ -252,5 +288,128 @@ pub(super) trait Inner: StateWrite {
     }
+
+    async fn _update_available_liquidity(
+        &mut self,
+        trading_pair: DirectedTradingPair,
+        position: &Position,
+        prev_position: &Option<Position>,
+    ) -> Result<()> {
+        tracing::debug!(?trading_pair, "updating available liquidity indices");
+
+        // Query the current available liquidity for this trading pair, or zero if the trading pair
+        // has no current liquidity.
+        let current_a_from_b = self
+            .nonverifiable_get_raw(&state_key::internal::routable_assets::a_from_b(
+                &trading_pair,
+            ))
+            .await?
+            .map(|bytes| {
+                Amount::from_be_bytes(
+                    bytes
+                        .try_into()
+                        .expect("invalid a_from_b stored in nonverifiable"),
+                )
+            })
+            .unwrap_or_default();
+
+        // Get the current reserves of this position, or zero if the position is newly added.
+        let current_reserves = prev_position
+            .as_ref()
+            .map(|prev| prev.reserves.clone())
+            .unwrap_or(Reserves {
+                r1: Amount::zero(),
+                r2: Amount::zero(),
+            });
+
+        // Use the current reserves to compute `current_position_contribution` (denominated in asset_1).
+        let current_position_contribution = match prev_position {
+            Some(prev_position) => {
+                // Return the amount of asset A purchasable with all reserves of asset B.
+                prev_position
+                    .phi
+                    .fill(
+                        Value {
+                            asset_id: trading_pair.end,
+                            amount: prev_position
+                                .reserves_for(trading_pair.end)
+                                .unwrap_or_default(),
+                        },
+                        &current_reserves,
+                    )?
+                    .2
+                    .amount
+            }
+            None => Amount::zero(),
+        };
+
+        // Use the new reserves to compute `new_position_contribution`,
+        // the amount of asset A purchasable with all reserves of asset B.
+        let new_position_contribution = match position.state {
+            position::State::Opened => {
+                // Return the amount of asset A purchasable with all reserves of asset B.
+                position
+                    .phi
+                    .fill(
+                        Value {
+                            asset_id: trading_pair.end,
+                            amount: position.reserves_for(trading_pair.end).unwrap_or_default(),
+                        },
+                        &position.reserves,
+                    )?
+                    .2
+                    .amount
+            }
+            _ => Amount::zero(),
```

Comment:
We might want to re-organize the method around a match statement over the previous position and the state of the current one. The state machine for this method looks like something like this:

```rust
match (current_position.state, prev_position) {
    (State::Opened, None) => we just have to add to the index, no deletion necessary
    (State::Opened, Some(prev)) => we add to the index, deleting previous entry
    (State::Closed, None) => we return early nothing to be done,
    (State::Closed, Some(prev)) => compute the previous contribution and erase it from the current index
}
```



I think we want to re-organize this method around a match statement on the current position's state, returning early if the current position is `Closed` for example.